src/main/rx/flysky.c

```diff
@@ -0,0 +1,496 @@
+
+                    if (sendTelemetry)
+                    {
+                        buildAndWriteTelemetry(packet);
+                        newState = A7105_TX;
+                        sendTelemetry = false;
+                    }
+
+                    result = RX_SPI_RECEIVED_DATA;
+                }
+
+                if (newState != A7105_TX)
+                {
+                    A7105WriteReg(A7105_0F_CHANNEL, getNextChannel(1));
+                }
+            }
+            break;
+
+        case FLYSKY_2A_PACKET_BIND1:
+        case FLYSKY_2A_PACKET_BIND2:
+            if (!binded)
+            {
+                resetTimeout(timeStamp);
+
+                flySky2ABindPkt_t *bindPacket = (flySky2ABindPkt_t*) packet;
+
+                if (bindPacket->rfChannelMap[0] != 0xFF)
+                {
+                    memcpy(rfChannelMap, bindPacket->rfChannelMap, NUMFREQ);       /* get TX channels */
+                }
+
+                txId = bindPacket->txId;
+                bindPacket->rxId = rxId;
+                memset(bindPacket->rfChannelMap, 0xFF, 26);                          /* erase channelMap and 10 bytes after it */
+
+                binded = ((bindPacket->state != 0) && (bindPacket->state != 1));
+                if (binded) { result = RX_SPI_RECEIVED_BIND; }                     /* bind complete */
+
+                A7105WriteFIFO(packet, FLYSKY_2A_PAYLOAD_SIZE);
+
+                newState = A7105_TX;
+            }
+            break;
+
+        default:
+            break;
+    }
+
+    A7105Strobe(newState);
+    return result;
+}
+
+
+static inline rx_spi_received_e flySkyReadAndProcess (uint8_t *payload, const uint32_t timeStamp)
+{
+    rx_spi_received_e result = RX_SPI_RECEIVED_NONE;
+    uint8_t packet[FLYSKY_PAYLOAD_SIZE];
+
+    uint8_t bytesToRead = (binded) ? (5 + 2*FLYSKY_CHANNEL_COUNT) : (5);
+    A7105ReadFIFO(packet, bytesToRead);
+
+    const flySkyRcDataPkt_t *rcPacket = (const flySkyRcDataPkt_t*) packet;
+
+    if (binded && rcPacket->type == FLYSKY_PACKET_RC_DATA && rcPacket->txId == txId)
+    {
+        checkRSSI();
+        resetTimeout(timeStamp);
+
+        if (payload) { memcpy(payload, rcPacket->data, 2*FLYSKY_CHANNEL_COUNT); }
+
+        A7105WriteReg(A7105_0F_CHANNEL, getNextChannel(1));
+
+        result = RX_SPI_RECEIVED_DATA;
+    }
+
+    if (!binded && rcPacket->type == FLYSKY_PACKET_BIND)
+    {
+        resetTimeout(timeStamp);
+
+        txId = rcPacket->txId;
+        flySkyCalculateRfChannels();
+
+        A7105WriteReg(A7105_0F_CHANNEL, getNextChannel(0));
+
+        binded = true;
+
+        result = RX_SPI_RECEIVED_BIND;
+    }
+
+    A7105Strobe(A7105_RX);
+    return result;
+}
+
+
+static inline bool isValidPacket (const uint8_t *packet)
+{
+    const flySky2ARcDataPkt_t *rcPacket = (const flySky2ARcDataPkt_t*) packet;
+    return (rcPacket->rxId == rxId && rcPacket->txId == txId);
+}
+
+
+static inline void buildAndWriteTelemetry (uint8_t *packet)

```

Comment:
FlySky / iBus telemetry is already implemented in https://github.com/vladisenko/betaflight/blob/4a0c89baf722135accf152712278db8ef7fc51c0/src/main/telemetry/ibus_shared.c. Instead of replicating the code, this should be using the existing implementation.

Reply:
The external protocol is described in the ibus_shared.c. I implement the internal protocol. Completely different implementations.

Reply:
This implementation looks reasonably similar to what is done in https://github.com/vladisenko/betaflight/blob/4a0c89baf722135accf152712278db8ef7fc51c0/src/main/telemetry/ibus_shared.c#L122, so you should be able to reuse this, by making `sendIbusMeasurement` virtual.
The advantage of this is that it eliminates duplication of data collection, and reduces the amount of work required for future maintenance of telemetry code.

For an example of how this can be done, see https://github.com/vladisenko/betaflight/blob/4a0c89baf722135accf152712278db8ef7fc51c0/src/main/telemetry/frsky.c#L474

Reply:
@mikeller , while the code in `ibus_shared.c` could be re-used, I don't think this is advisable in its current form - it is too tightly bound to writing to serial. It would be better to re-factor the ibus telemetry code to use sbufs, like the CRSF code does, and then re-use that. I think that refactoring work should be done after this code is merged in.

Reply:
@martinbudden: Potentially. But since this is slated for 3.3, there will be some delay until it gets merged, so the refactoring can be done before or after the branch.